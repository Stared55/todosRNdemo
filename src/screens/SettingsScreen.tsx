import React, { useEffect } from 'react';
import { Alert, Platform } from 'react-native';
import Share from 'react-native-share';
import { useDispatch, useSelector } from 'react-redux';
import Navbar from 'src/components/navigation/Navbar';
import CustomButton from 'src/components/shared/CustomButton';
import ScrollViewWrapper from 'src/components/shared/ScrollViewWrapper';
import { AppState } from 'src/store/store';
import { clearTodos, getTodos, saveTodos } from 'src/store/todos/todosActions';
import { Colors } from 'src/utils/styles/colors';
import { xScale, yScale } from 'src/utils/styles/scale';
import { BottomTabNavigationProp } from '@react-navigation/bottom-tabs';
import { TabParamList } from 'src/components/navigation/RootNavigator';

type SettingsScreenNavigationProp = BottomTabNavigationProp<
  TabParamList,
  'Settings'
>;

export interface SettingsScreenProps {
  navigation?: SettingsScreenNavigationProp;
}

const SettingsScreen: React.SFC<SettingsScreenProps> = () => {
  const { todos } = useSelector((state: AppState) => state.todos);
  const [message, setMessage] = React.useState<string>('');
  const dispatch = useDispatch();

  useEffect(() => {
    createPlainText();
  }, [todos]);

  const createPlainText = () => {
    let plainText = '';
    for (let i = 0; i < todos.length; i++) {
      const todo = todos[i];
      const item = `${todo.title} ${todo.completed ? '+' : '-'}\n`;
      plainText = plainText.concat(item, plainText);
    }
    setMessage(plainText);
  };

  const options: any = Platform.select({
    ios: {
      activityItemSources: [
        {
          placeholderItem: { type: 'text', content: message },
          item: {
            default: { type: 'text', content: message },
            message: message,
          },
        },
      ],
    },
    default: {
      title: 'My todos',
      subject: 'My todos',
      message,
    },
  });

  const handleSaveTodos = () => {
    dispatch(saveTodos());
  };

  const handleLoadTodos = () => {
    dispatch(getTodos());
  };

  const handleClearTodos = () => {
    dispatch(clearTodos());
  };

  const handleShareList = () => {
    Share.open(options)
      .then((res) => {
        console.log(res);
      })
      .catch((err) => {
        err && console.log(err);
      });
  };

  const warnAlert = () => {
    Alert.alert(
      'Warning',
      'Are you sure you want to clear your data?',
      [
        {
          text: 'Clear',
          onPress: () => handleClearTodos(),
          style: 'destructive',
        },
        {
          text: 'Cancel',
          onPress: () => null,
          style: 'cancel',
        },
      ],
      { cancelable: false },
    );
  };

  return (
    <>
      <Navbar title="Settings" />
      <ScrollViewWrapper
        contentContainer={{
          justifyContent: 'center',
          flex: 1,
          paddingHorizontal: xScale(20),
        }}>
        <CustomButton label="Save Todos" handler={handleSaveTodos} />
        <CustomButton label="Load Todos" handler={handleLoadTodos} />
        <CustomButton
          label="Clear Todos + AsyncStorage"
          handler={warnAlert}
          containerStyles={{ backgroundColor: Colors.darkRed }}
        />
        <CustomButton
          label="Share List"
          handler={handleShareList}
          containerStyles={{
            marginTop: yScale(46),
          }}
        />
      </ScrollViewWrapper>
    </>
  );
};

export default SettingsScreen;
