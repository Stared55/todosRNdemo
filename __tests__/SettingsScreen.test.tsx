import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import SettingsScreen from 'src/screens/SettingsScreen';

jest.mock('react-redux', () => ({
  useDispatch: () => () => jest.fn(),
  useSelector: (selector: any) =>
    selector({
      loading: false,
      todos: [],
      error: null,
    }),
}));

describe('SettingsScreen', () => {
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    component = shallow(<SettingsScreen />);
    props = component.props();
    jest.clearAllMocks();
  });

  it('should match to snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render navbar with correct label', () => {
    const title = props.children[0].props.title;
    expect(title).toEqual('Settings');
  });

  it('should render 4 buttons', () => {
    const buttons = props.children[1].props.children;
    expect(buttons).toHaveLength(4);
  });
});
