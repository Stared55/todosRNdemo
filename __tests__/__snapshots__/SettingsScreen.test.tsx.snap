// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen should match to snapshot 1`] = `
<Fragment>
  <Navbar
    title="Settings"
  />
  <ScrollViewWrapper
    contentContainer={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 42.857142857142854,
      }
    }
  >
    <CustomButton
      handler={[Function]}
      label="Save Todos"
    />
    <CustomButton
      handler={[Function]}
      label="Load Todos"
    />
    <CustomButton
      containerStyles={
        Object {
          "backgroundColor": "#B93333",
        }
      }
      handler={[Function]}
      label="Clear Todos + AsyncStorage"
    />
    <CustomButton
      containerStyles={
        Object {
          "marginTop": 90.24117647058823,
        }
      }
      handler={[Function]}
      label="Share List"
    />
  </ScrollViewWrapper>
</Fragment>
`;
